Add history_stats sensor

bokub opened the pull request:
**Add new sensor/history_stats component**


**Feature request**: [History statistics](https://community.home-assistant.io/t/history-statistics-component/10194)

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#1901

**Example entry for `configuration.yaml` (if applicable):**
```yaml
sensor:
  - platform: history_stats
    name: Lamp ON today
    entity_id: light.my_lamp
    state: 'on'
    start: {{ _TODAY_ }}
    end: {{ _NOW_ }}
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51

bokub committed:
Add new sensor/history_stats component

bokub committed:
Add real unit tests

fabaff commented on homeassistant/components/sensor/history_stats.py:
```diff
@@ -0,0 +1,303 @@
+    def pretty_datetime(timestamp):
+        """Format a timestamp to a formatted datetime in the local timezone."""
+        return dt_util.as_local(dt_util.utc_from_timestamp(timestamp)) \
+            .strftime('%Y/%m/%d %H:%M:%S')
```
This would introduce an addition time format. We are trying to use ISO 8601 everything. Also, Europeans use a different format (at least I do anyway) ;-)

bokub replied:
Well, according to wikipedia, an ISO 8601 datetime would be something like `2016-12-13T14:23:09Z`.

I'm surprised as I've never seen a datetime formatted like that in home assistant, even in the history component : 

----------------
![history](https://cloud.githubusercontent.com/assets/17952318/22404700/a92a7ec6-e636-11e6-91b8-618ebcee33fd.PNG)

----------------

Wouldn't something like `2016-12-13 14:23:09` be better ?

fabaff replied:
Sorry, I'm only talking about the backend.

```bash
...type=upper, sensor_value=18.0 @ 2017-01-30T15:27:37.559413+01:00>>
```

fabaff reviewed (dismissed):
Just some quick comments for now as `history` and `recorder` is not really my area of expertise.

bokub commented on homeassistant/components/sensor/history_stats.py:
```diff
@@ -0,0 +1,303 @@
+    vol.Required(CONF_ENTITY_ID): cv.entity_id,
+    vol.Required(CONF_STATE): cv.slug,
+    vol.Optional(CONF_START, default=None): cv.template,
+    vol.Optional(CONF_END, default=None): cv.template,
```
Because you might want to provide only `start` and `duration`. In that case, `end` should stay undefined.

fabaff replied:
This was unclear because the `voluptuous` was missing.

bokub commented on homeassistant/components/sensor/history_stats.py:
```diff
@@ -0,0 +1,303 @@
+        return True
+
+    @property
+    def state_attributes(self):
```
I'll change that, but just to let you know, the method is called `state_attributes` in `sensor/statisctics.py`. 

I used it to help me code my component.

fabaff replied:
So you can blame me that I did it wrong in the first place.

bokub commented on homeassistant/components/sensor/history_stats.py:
```diff
@@ -0,0 +1,303 @@
+            .replace("_TODAY_", "_THIS_HOUR_.replace(hour=0)")
+            .replace("_THIS_HOUR_", "_THIS_MINUTE_.replace(minute=0)")
+            .replace("_THIS_MINUTE_", "_NOW_.replace(second=0)")) \
+            .replace("_NOW_", "now()")
```
All right, I could have done that, but I wasn't sure it would mix well with what exists right now.

What do you think about my aliases format ? Should I add them like that in `helpers.template` so they can work everywhere ?

fabaff replied:
We have `now()` and `utcnow()` plus some helpers.  As a non-native English speaker I would be fine with something like `one_week()` or `this_week()/current_week()`.

I would like to suggest that you create an new PR with the changes for the templates. This way we can keep thing separated and it would be easier to get feedback.

bokub replied:
No problem about adding those helpers in a future PR, however those helpers will refer to datetimes, so users will have to add `as_timestamp(...)` around everything. It might no be ideal, but still helpful.

And what to do with the current PR ? Should I remove my aliases ? Or should I leave them here until the templates can handle them ?

fabaff replied:
The proper way would be to make the templates and then the update this PR. The issue here is that if we are going to remove the templates later from the sensor we will most likely end-up with a breaking change .

homeassistant commented:
Hi @bokub, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5610).

Once you do that we will be able to review and accept this pull request.

Thanks!

homeassistant commented:
Hi @bokub, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5610).

Once you do that we will be able to review and accept this pull request.

Thanks!

homeassistant commented:
Hi @bokub, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5610).

Once you do that we will be able to review and accept this pull request.

Thanks!

fabaff commented on homeassistant/components/sensor/history_stats.py:
```diff
@@ -0,0 +1,304 @@
+    if params != 2:
+        _LOGGER.error('You must provide exactly 2 of the following:'
+                      + ' start, end, duration')
+        return False
```
This could be handled by `voluptuous` with `vol.Exclusive` and `vol.Inclusive`.

bokub replied:
`vol.Inclusive` is good when you want either 3/3 or 0/3 parameters.
`vol.Exclusive` is good when you want either 1/3 or 0/3 parameters.
What I want is exactly 2/3 of them, not more, not less. That means my schema can be one of 3 possible schemas.

I have been trying to do combinations of these for the last 2 hours but I really cannot make it, I'm really confused by how voluptuous works, I'm sorry.

fabaff replied:
Not all combinations can be checked. I thought that it would work as the AWS notification platforms are doing something similar.

bokub committed:
Code style changes requested for pull request

bokub commented:
Hi @fabaff ,
There are still 2 things unclear about my pull request, could you please tell me what you think when you have a minute ?

- What should I do with my current aliases ? Should I leave them in my code until the templates can handle them, or should I just remove them ?

- About the voluptuous validation, I'm really stuck as I cannot validate _exactly 2 out of 3_ (see [my previous comment](https://github.com/home-assistant/home-assistant/pull/5610#discussion_r98519812)). I Asked for help on stackoverflow, but nobody seem to be able to help me.

Thanks !

roflmao commented:
Is there any possibility to count number of on/off states? I use a fibaro wallplug on my washing machine and it would be nice to have a counter on how many "on" states it has done ex the last week.

bokub commented:
@roflmao sure, this is no more than 5 lines of code to add. However, I'd prefer to see my pull request merged before submitting new features. 
If you want, you can post your feature request as an issue on [my repo](https://github.com/bokub/ha-history-stats), with details about what exactly you want to be counted, as well as what you would write in your `configuration.yaml`

fabaff commented:
> What should I do with my current aliases ? Should I leave them in my code until the templates can handle them, or should I just remove them ?

See comment above. I would prefer if we don't include them to remove a short time later.

> About the voluptuous validation, I'm really stuck as I cannot validate exactly 2 out of 3 (see my previous comment). I Asked for help on stackoverflow, but nobody seem to be able to help me.

If you want full flexibility and need the three options, the only way would be to create a new rule but this seems a bit over-kill. Let's stick with the current solution till somebody come up with something better.

balloob commented:
To add a voluptuous rule that validates 2 out of 3 are there, just move that check into a function.

```python
def validate_my_two_things(value):  # Needs a better name obv.
    some_check = CONF_BLA in value
    other_check = CONF_BLUB in value

    if some_check but not other_check:
        raise vol.Invalid('You need BLA if you don't have BLUB")

    return value

PLATFORM_SCHEMA = vol.All(vol.Schema({
    …
}), validate_my_two_things)
```

bokub committed:
Remove time aliases & allow datetime values

bokub committed:
Reformat unit tests

bokub commented:
Hi again @fabaff, I have done the changes you requested : 

- All the configuration validation is made with voluptuous (thanks again @balloob)
- I have removed the _time aliases_ feature

In addition, I have made a little improvement : `start` and `end` can now be datetimes **or** timestamps. 
This allows both `{{ now() }}` and `{{ as_timestamp(now()) - 3600 }}` to work with my component, making it more flexible.

Of course, I have updated the documentation too,

balloob commented on homeassistant/components/sensor/history_stats.py:
```diff
@@ -0,0 +1,314 @@
+        # Without this method, the recorder does not start its connection
+        # itself, resulting in an infinite loop blocking the boot of home
+        # assistant. It may be a nasty bug.
+        if recorder._INSTANCE.db_ready._flag:
```
Do not do this. That's an internal method.

balloob commented on homeassistant/components/sensor/history_stats.py:
```diff
@@ -0,0 +1,314 @@
+        # assistant. It may be a nasty bug.
+        if recorder._INSTANCE.db_ready._flag:
+            return True
+        time.sleep(0.5)  # Wait, just in case db is starting
```
No sleeps please.

balloob commented:
Please remove all the sleeps from the code.

All test logs always contain the slowest 10 tests, you've made it to the top:

![screen shot 2017-02-07 at 22 14 41](https://cloud.githubusercontent.com/assets/1444314/22725553/077223b4-ed83-11e6-98ca-f51491714684.png)

balloob commented on homeassistant/components/sensor/history_stats.py:
```diff
@@ -0,0 +1,314 @@
+        start = dt_util.as_utc(start)
+        end = dt_util.as_utc(end)
+
+        if not HistoryStatsHelper.wait_till_db_ready():
```
You cannot do sleeps in the event loop

balloob commented on homeassistant/components/sensor/history_stats.py:
```diff
@@ -0,0 +1,314 @@
+            return
+
+        # Get history between start and end
+        history_list = history.state_changes_during_period(
```
Querying the history is I/O which is not allowed in the event loop.

balloob commented on homeassistant/components/sensor/history_stats.py:
```diff
@@ -0,0 +1,314 @@
+        # Save value in hours
+        self.value = elapsed / 3600
+
+    def update_period(self):
```
If this is an async method, mark it as such with either `@callback` or `@asyncio.coroutine`. Also prefix the name with `async_`

bokub commented:
Hi @balloob, thanks for your code review. I feel like almost all of the mistakes you've noticed are because I made a history query in `async_update`.

>  Querying the history is I/O which is not allowed in the event loop.


I perfectly know that the `sleep()` and the internal method calls are crappy, but they're the best solution to a problem I encountered : home assistant blocking during the boot. I posted a topic in the home assistant community but nobody seemed to be able to explain. I guess that's because I shouldn't have called the history in `async_update`.

I have to be honest, my knowledge of python and home assistant are not strong enough, I don't even know how I should have done it, or how to correct it now, and I'm really sorry about that.

balloob commented:
I think the first step will be to not use any async functions. We have a fully threaded API. The second will be to not use sleep.

Don't worry that you are not too familiar with Python or Home Assistant. Let's see how far we can get together 👍 

- Rebase on latest dev
- remove `@asyncio.coroutine` and `@callback` everywhere
- Use `setup_platform`, rename `async_add_devices` to `add_devices` and call it as a normal function instead of a `yield from`, remove the second argument (`True`) to `add_devices`
- add a new listener in your constructor
```python
hass.bus.listen_once(
    EVENT_HOMEASSISTANT_START,
    lambda event: self.schedule_update_ha_state(True))
```
 - Change your state change listener to:
```python
track_state_change(
    hass, entity_id,
    lambda event: self.schedule_update_ha_state(True))
```
 - rename `async_update` to `update`
 - for templates, call `render` instead of `async_render`
 - remove `wait_till_db_ready` 

I can't guarantee that all will work when you make these changes, but you should get pretty far

balloob commented:
Also for the config:
- Let's simplify it a bit for the first version.
- Remove `end`. Only allow template for `start`
- For duration, there is a config validation called `cv.period`. It will not allow templates but will allow [different ways of specifying a duration](https://home-assistant.io/getting-started/scripts/#delay) (but not the template example).

bokub committed:
Remove all async behavior in history_stats

bokub commented:
Oh thank you so much for your help ! I removed all async behavior from my component and now it finally works like a charm ! 👍

The only thing I kind of disagree with is what you want me to do with the config:
> Remove end. Only allow template for start

Aren't you afraid that it will make templates way more difficult to write ? For example, for the period `yesterday`, the following code:

    end: '{{ now().replace(hour=0).replace(minute=0).replace(second=0) }}'
    duration: '24:00:00'

becomes:

    start: '{{ as_timestamp(now().replace(hour=0).replace(minute=0).replace(second=0)) - 86400 }}'
    duration: '24:00:00'

...which implies converting a date to a timestamp in order to subtract 24 hours !

By the way, the validation now works flawlessly, and I'd say my documentation is pretty clear about how to handle *start*, *end*, and *duration*.

What do you think about it ?

bokub committed:
Change duration format

balloob reviewed (approved):
Looks great 🐬
